File: tests/e2e/explorer_rename_test.go
package e2e

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/eugenioenko/ttt/internal/widgets"

	"github.com/gdamore/tcell/v3"
)

// renameVia drives the explorer rename dialog for the node at path, replacing
// its name with newName.
func (h *testHarness) renameVia(path, newName string) {
	h.t.Helper()
	h.app.ExplorerContextNode = &widgets.TreeNode{ID: path}
	h.exec("explorer.rename")
	h.redraw()

	// The dialog is pre-filled with the current basename and places no
	// selection, so clear it before typing the replacement.
	for range filepath.Base(path) {
		h.pressKey(tcell.KeyBackspace, tcell.ModNone)
	}
	h.typeText(newName)
	h.pressKey(tcell.KeyEnter, tcell.ModNone)
	h.redraw()
}

// Renaming a file in the explorer used to leave the open tab pointing at the
// old path, so Ctrl+S wrote the buffer back out under the name the file had
// been renamed away from, leaving a ghost duplicate on disk. Issue #284.
func TestExplorerRenameUpdatesOpenTabAndSavesToNewPath(t *testing.T) {
	h := newTestHarness(t, 100, 30)
	defer h.stop()

	old := filepath.Join(h.dir, "alpha.txt")
	h.app.EditorGroup.OpenFile(old)
	h.redraw()

	h.renameVia(old, "renamed.txt")

	newPath := filepath.Join(h.dir, "renamed.txt")
	if got := h.app.EditorGroup.ActiveFilePath(); got != newPath {
		t.Fatalf("tab still points at the old path: got %q, want %q", got, newPath)
	}

	h.app.EditorGroup.ActiveBuffer().Lines = []string{"edited"}
	h.app.EditorGroup.ActiveBuffer().Dirty = true
	h.exec("file.save")

	if _, err := os.Stat(old); !os.IsNotExist(err) {
		t.Fatalf("save recreated %s as a ghost duplicate", old)
	}
	data, err := os.ReadFile(newPath)
	if err != nil {
		t.Fatalf("read renamed file: %v", err)
	}
	if got := string(data); got != "edited\n" && got != "edited" {
		t.Fatalf("renamed file contents: got %q, want edited", got)
	}
}

// A folder rename has to carry every tab beneath it, not just an exact match.
func TestExplorerRenameFolderUpdatesNestedTab(t *testing.T) {
	h := newTestHarness(t, 100, 30)
	defer h.stop()

	nested := filepath.Join(h.dir, "subdir", "nested.txt")
	h.app.EditorGroup.OpenFile(nested)
	h.redraw()

	h.renameVia(filepath.Join(h.dir, "subdir"), "renamed-dir")

	want := filepath.Join(h.dir, "renamed-dir", "nested.txt")
	if got := h.app.EditorGroup.ActiveFilePath(); got != want {
		t.Fatalf("nested tab path: got %q, want %q", got, want)
	}
}

// A file that is not open must still rename cleanly, and must not disturb the
// tab that is open.
func TestExplorerRenameLeavesOtherTabsAlone(t *testing.T) {
	h := newTestHarness(t, 100, 30)
	defer h.stop()

	openPath := filepath.Join(h.dir, "alpha.txt")
	h.app.EditorGroup.OpenFile(openPath)
	h.redraw()

	h.renameVia(filepath.Join(h.dir, "beta.txt"), "beta-renamed.txt")

	if got := h.app.EditorGroup.ActiveFilePath(); got != openPath {
		t.Fatalf("unrelated rename moved the open tab: got %q, want %q", got, openPath)
	}
	if _, err := os.Stat(filepath.Join(h.dir, "beta-renamed.txt")); err != nil {
		t.Fatalf("expected beta-renamed.txt on disk: %v", err)
	}
}

File: tests/functional/explorer-rename.test.js
import { describe, it, expect, afterEach } from "vitest";
import * as tui from "./tui.js";
import {
  createTempDir,
  createTempFile,
  cleanupDir,
  fileExists,
  readFile,
} from "./helpers.js";
import { writeFileSync } from "node:fs";
import { join } from "node:path";

let dir;

afterEach(() => {
  tui.kill();
  if (dir) cleanupDir(dir);
});

// The rename dialog is pre-filled with the current basename and places no
// selection, so the old name has to be cleared before typing the new one.
function renameSelectedTo(oldName, newName) {
  tui.exec("Explorer: Rename");
  for (let i = 0; i < oldName.length; i++) tui.press("backspace");
  tui.type(newName);
  tui.press("enter");
}

describe("explorer rename", () => {
  it("should retitle the open tab when its file is renamed", () => {
    dir = createTempDir();
    createTempFile(dir, "alpha.txt", "ALPHA CONTENT");

    tui.start(dir);
    tui.waitFor("Explore");

    tui.press("ctrl+0");
    tui.press("arrow_down");
    tui.press("enter");

    renameSelectedTo("alpha.txt", "renamed.txt");

    const s0 = tui.snapshot();
    const { snapshots } = tui.run();

    expect(snapshots[s0]).toContain("renamed.txt");
    expect(snapshots[s0]).not.toContain("alpha.txt");
  });

  // Issue #284: the tab kept the old path, so saving wrote the buffer back out
  // under the name the file had been renamed away from.
  it("should save to the new path without recreating the old file", () => {
    dir = createTempDir();
    createTempFile(dir, "alpha.txt", "ALPHA CONTENT");

    tui.start(dir);
    tui.waitFor("Explore");

    tui.press("ctrl+0");
    tui.press("arrow_down");
    tui.press("enter");

    renameSelectedTo("alpha.txt", "renamed.txt");

    tui.type("EDITED ");
    tui.exec("Save File");
    tui.run();

    expect(fileExists(join(dir, "alpha.txt"))).toBe(false);
    expect(fileExists(join(dir, "renamed.txt"))).toBe(true);
    expect(readFile(join(dir, "renamed.txt"))).toContain("EDITED");
  });

  it("should keep editing the same buffer after a rename", () => {
    dir = createTempDir();
    createTempFile(dir, "alpha.txt", "ORIGINAL LINE");

    tui.start(dir);
    tui.waitFor("Explore");

    tui.press("ctrl+0");
    tui.press("arrow_down");
    tui.press("enter");

    renameSelectedTo("alpha.txt", "renamed.txt");

    const s0 = tui.snapshot();
    const { snapshots } = tui.run();

    expect(snapshots[s0]).toContain("ORIGINAL LINE");
  });
});

describe("explorer refresh", () => {
  it("should pick up a file created outside the editor when r is pressed", () => {
    dir = createTempDir();
    createTempFile(dir, "alpha.txt", "alpha");

    tui.start(dir);
    tui.waitFor("Explore");

    tui.press("ctrl+0");

    // The explorer has already listed the directory; add a file behind its back.
    writeFileSync(join(dir, "created-outside.txt"), "outside");

    tui.press("r");

    const s0 = tui.snapshot();
    const { snapshots } = tui.run();

    expect(snapshots[s0]).toContain("created-outside.txt");
  });
});
